### ISSUE TITLE
Support for building without vendored dependencies

### ISSUE THREAD

**Issue by NexAdn** [brainboxdotcc/DPP#207]
**Is your feature request related to a problem? Please describe.**
Currently, DPP ships with three dependencies directly integrated into the source tree.
While this is very comfortable for users and developers, it is a nightmare for package maintainers, trying to integrate the software properly into their system and (in the process) removing vendored dependencies.

I managed to remove nlohmann JSON as a vendored dependency with a bunch of `sed` and `rm`, but creating this patch was already painful.

From the point of properly documenting all required licenses and being able to install patched versions of dependencies without having to depend on the downstream library to bump a new release, it is desirable for package maintainers to have the libary only use system-installed dependencies and no vendored ones.

**Describe the solution you'd like**
Preferably, removing all vendored dependencies.
But since I see valid reasons for vendoring dependencies which aren't commonly available in distributions, it's probably better to move the vendored dependencies to a separate subdirectory and providing a CMake option to disable the inclusion of vendored dependencies.

**Describe alternatives you've considered**
The alternative is the status quo, but I've explained why I'd like to see the change.

**Additional context**

* The collection of all the patches I've written to package DPP on my machine and make the integration in other projects work: https://gitlab.com/NexAdn/dpp-gentoo-patches

**Comment by braindigitalis** [comment]
The three dependencies that are vendored by DPP are specific versions, and you can't just build it with any version of cpp-httplib, nlohmann::json or fmtlib.

Also, two of the three are modified to remove their header-only nature. Drop-in replacements from the system are not valid replacements and would be detrimental to the build speed and compiled size of DPP based bots.

Is there a specific reason to not need to vendor these header-only libraries in (this is the intent of header only libraries after all) - the changes your patches make are kind of wide-reaching. Is there a reason for example for not running ldconfig? (this is needed, or the library cannot be found) or for changing the names of all the variables in the cmake?

**Comment by braindigitalis** [comment]
looking at the changes, your intent is to get the lib into gentoo emerge? Would this be something people would be interested in?

**Comment by NexAdn** [comment]
> The three dependencies that are vendored by DPP are specific versions, and you can't just build it with any version of cpp-httplib, nlohmann::json or fmtlib.

Actually I can (at least I could for nlohmann json). If I can't, please tell me, _why_ I can't. Because, if it is technically possible (although undesired) to build without vendored dependency, I _can_ build without vendored dependencies.

> Also, two of the three are modified to remove their header-only nature. Drop-in replacements from the system are not valid replacements and would be detrimental to the build speed and compiled size of DPP based bots.

Which of those dependencies have which modifications? I didn't see any documentation on this.
Actually, there isn't even documentation on which files are from which dependency and how they are integrated (although the latter is pretty obvious for header-only libraries).
If the modifications are only to eliminate the header-only nature, building with the normal libraries should still be possible if the includes are adjusted properly.

> detrimental to the build speed and compiled size of DPP based bots.

Well, the build speed is an argument for removing the header-only nature. But in my opinion, this does _not_ justify maintaining a fork – even if upstream refuses offering a non-header-only version of the library (which I'd prefer, too, btw).

> Is there a specific reason to not need to vendor these header-only libraries in (this is the intent of header only libraries after all) - the changes your patches make are kind of wide-reaching.

There are two reasons, which I've already stated. Bumping versions always depends on a package maintainer. Especially if security patches happen in upstream library versions, the package maintainer's hands are tied if the downstream software maintainer (in this case, the DPP team) doesn't update their library soon. While some projects work around the issue for pinned dependencies by using bots that bump those pinned dependencies' versions every now and then, this isn't really possible to do automatically for vendored dependencies with custom patches – especially if those patches don't happen inside a forked repository.

I have seen people forking libraries to add their own patches and then forgetting to rebase the patches onto newer versions of the upstream library. And after a while it didn't seem so problematic until the fork was behind upstream over more than a year. And I've seen that for a networking library–so something security-critical! I don't trust anyone promising to keep their fork up to date. Some time, some changes upstream might make rebasing the patches too hard and it seems way too easy to just stick with the old library version.

> this is the intent of header only libraries after all

I never said that this was a good intent.
I am against writing header-only libraries.

> the changes your patches make are kind of wide-reaching.

Well, not all patches are relevant to this issue. Relevant to this issue is `dpp-remove-vendored-nlohmann-json.patch` which is just the result of a few `sed` operations to replace all inclusions of nlohmann json and to remove the nlohmann json files from the source tree.

> Is there a reason for example for not running ldconfig?

First of all, that's not relevant to this issue.
Second, I did this because of a sandbox violation.
Running ldconfig manually should not be necessary.
You don't have an SONAME specified for the library and it gets installed as `libdpp.so` anyway (and not something like `libdpp.so.1.1`).
Thus, the library _will_ be found when linking with `-ldpp`.
ldconfig is required if you have installed DPP as `libdpp.so`, but with SONAME `libdpp.so.x`, because now ELFs linking agains your library would link against `libdpp.so.x` and so the symlink `libdpp.so.x -> libdpp.so` must exist for the ELFs to work.
However, you don't have an SONAME (or rather, the SONAME `libdpp.so`), so the libraries will link against `libdpp.so`. Since this is the same name as the installed .so file, everything will work out without running ldconfig.

Apart from that, I trust the Gentoo developers that running ldconfig should be prohibited in the install part of the program and thus blocked by the sandbox. I've never had such a sandbox violation like this, even with more complex libaries that actually had versioned SONAMEs.

Also, if you decide to use a versioned SONAME, CMake should be able to handle the symlinking done by ldconfig itself.

However, all this applies only if DPP is installed to standard library paths (which I ensured with `dpp-fix-install.patch`). I don't know how linking against DPP without ldconfig behaves if installed to a non-standard library path.

> for changing the names of all the variables in the cmake?

Well, that was just my attempt at making the installed DPP work. The default installation just failed to even detect DPP properly, since some CMake files were installed as `libdpp*.cmake` and some as `dpp*.cmake`. But I haven't checked if this is Gentoo-related or not, yet.

Also, I am not sure if the patches modify the files in a way that are compatible with DPP builds on other systems (I haven't compared with the prebuilt packages, yet).

> looking at the changes, your intent is to get the lib into gentoo emerge?

Since I run my bot on a Gentoo server, yes.
At least I intend having it packaged in a private overlay (which I've already done, btw).

> Would this be something people would be interested in?

From the perspective of someone who just started using this library (me), I really like the library.
That's after all, why I decided to migrate my bot to this library.

Discord libraries aren't so popular as to make it to the “big” distro's repos (I don't intend on packaging DPP for the main portage tree, since Gentoo's proxy maintainers don't have the time to proxy-maintain new small packages), but an easily packageable library allows at least for integration in user-maintained repositories like the AUR or Gentoo's GURU (which I might upload my package to as soon as I manage to package DPP properly).

Really, I don't have a problem with you using vendored dependencies, as long as I have an option not to use them.
At least for fmt and nlohmann json we can assume that their API should be pretty stable across version bumps and thus the extra support overhead will be low to none.
For the httplib, if there are frequent breaks, you might get discouraged to update the vendored dependency regularly and thus there is a security risk when using the vendored dependency. In that case, vendoring that dependency is a bad idea to begin with, considering possible security problems.

**Comment by braindigitalis** [comment]
Pretty sure the majority of the issues you report are related to how youre using the library.
Most people do just follow our build instructions, and it works, on gentoo and elsewhere. If you do think these changes of yours are beneficial for the library as a whole (not just gentoo) then please do submit a pr.

I wouldnt merge in a PR that lets people chop and change their version of nlohmann or other vendored libs though though without some restriction on the version number. We cannot support, or test, every possible version of nlohmann people may try and bundle into their installation. It's pretty much standardized and we know what we have works, passes our unit and regression tests, we can't say the same for any random newer or older version. Same goes for fmtlib and httplib. This is why they are vendored and you'll see it's common practice amongst discord libs to vendor stuff like this (see aegis, sleepy discord, etc which bundle things like websocketpp and asio). 

As you can see by looking at fmtlib, it's split usually using its own splitter script, which turns it into cpp and hpp. Without these, the build times and binary sizes are significantly larger, especially when referring to fmtlib.

Also, please be aware that if you are intending to package this into an ebuild or similar, we do not guarantee, or ever plan to guarantee binary compatibility between builds. This is because Discord constantly incrementally change their API monthly resulting in binary breakage even between minor versions. If this is important, this is something we cannot do anything about nor do any discord libraries (which is likely why youve never seen any other Discord API wrappers for C++ included in a linux distro package manager)

**Comment by SirObby** [comment]
I don't see a need for D++ to be in portage (emerge). It doesn't seem necessary.

**Comment by braindigitalis** [comment]
thanks for the contribution, i have taken some of your patches and merged them into the master branch, such as the ability to turn off ldconfig and voice building, and setting the soversion symlinks.
I have not allowed for the user to choose their own fmtlib or nlohmann::json at this time.
As i have used your patches in part, if you do want the contributor role on the server please let me know (if you are on the server) at discord.gg/dpp
